nomt/src/cursor.rs

```diff
@@ -13,31 +14,60 @@ use nomt_core::{
 /// The number of items we want to request in a single batch.
 const PREFETCH_N: usize = 7;
 
+/// A cursor wrapping a [`PageCache`].
+///
+/// This performs I/O internally.
 pub struct PageCacheCursor {
+    pages: PageCache,
     root: Node,
     path: KeyPath,
     depth: u8,
-    /// The cached page and it's ID.
-    ///
-    /// Justification for the interior mutability is to use &self in [`Self::node`].
-    cached_page: RefCell<Option<(PageId, Page)>>,
+    // Invariant: this is always set when not at the root and never
+    // set at the root.
+    cached_page: Option<(PageId, Page)>,
 }
 
 impl PageCacheCursor {
     /// Creates the cursor pointing at the given root of the trie.
-    pub fn at_root(root: Node) -> Self {
+    pub fn at_root(root: Node, pages: PageCache) -> Self {
         Self {
             root,
             path: KeyPath::default(),
             depth: 0,
-            cached_page: RefCell::new(None),
+            pages,
+            cached_page: None,
         }
     }
 
     /// Moves the cursor back to the root.
-    pub fn reset(&mut self) {
+    pub fn rewind(&mut self) {
         self.depth = 0;
-        self.path = KeyPath::default();
+        self.cached_page = None;
+    }
+
+    /// Get the current position of the cursor expressed as a bit-path and length. Bits after the
+    /// length are irrelevant.
+    pub fn position(&self) -> (KeyPath, u8) {
+        (self.path, self.depth)
+    }
+
+    /// Jump to the node at the given path. Only the first `depth` bits are relevant.
+    /// It is possible to jump out of bounds, that is, to a node whose parent is a terminal.
+    pub fn jump(&mut self, path: KeyPath, depth: u8) {
+        self.path = path;
+        self.depth = depth;
+
+        if depth == 0 {
+            self.rewind();
+            return;
+        }
+
+        let n_pages = self.depth as usize / DEPTH;
+        let page_id = PageIdsIterator::new(&self.path)
+            .nth(n_pages)
+            .expect("all keys with <= 256 bits have pages; qed");
```

Comment:
I don't think the proof is adequate tbh. 

`nth` returning None/Some depends on the iterator behavior defined by `PageIdsIterator`. It is not documented, but we can assume it chops 256 by DEPTH-sized chunks, which gives 43. So as long as n_pages < 43 we should be fine. `self.depth` is an u8 and thus maximum is 255 and 255 / DEPTH = 42.